b3sum/src/main.rs

```diff
@@ -496,26 +510,43 @@ fn parse_check_line(mut line: &str) -> Result<ParsedCheckLine> {
 }
 
 fn hash_one_input(path: &Path, args: &Args) -> Result<()> {
-    let mut input = Input::open(path, args)?;
-    let output = input.hash(args)?;
-    if args.raw() {
-        write_raw_output(output, args)?;
-        return Ok(());
-    }
-    if args.no_names() {
+    let md = metadata(path).unwrap();
+    if md.is_dir() && args.recurse() {
```

Comment:
Have you thought about what will happen for directory entries that are symlinks in this case? It looks like if they're symlinks to directories, then the'll be opened as though they were files, which will probably cause an error. But in fixing this, we have to be careful not to infinitely loop on circular symlinks. The right thing to do here isn't clear to me, and we might want to look at what similar recursive tools do.

Reply:
No, good point, I hadn't considered that. My use-case is on Windows where symlinks are very unlikely to exist but I'll add another option `--follow-symlinks` to follow them, and investigate how to detect loops. IMHO it would make sense not to follow them by default but I don't mind doing the opposite and having a `--no-follow-symlinks` option.

Reply:
I think we need something similar to Javascript `WeakSet` to track symlinks